src/qrexec-agent/qrexec-agent.c

```diff
@@ -70,6 +70,95 @@ void DumpRequestList(void)
 }
 #endif
 
+struct _connection_info
+{
+    HANDLE handle;
+    int connect_domain;
+    int connect_port;
+};
+
+#define MAX_FDS 256
+static struct _connection_info connection_info[MAX_FDS];
+
+static void register_vchan_connection(HANDLE handle, int domain, int port)
+{
+    for (int i = 0; i < MAX_FDS; i++)
+    {
+        if (connection_info[i].handle == NULL)
+        {
+            EnterCriticalSection(&g_DaemonCriticalSection);
```

Comment:
The whole loop should be under the lock. Otherwise two threads could write to the same slot.